### ISSUE TITLE
State not propagated on invoke in multiple peer setup - noops

### ISSUE THREAD

**Issue by mrshah-at-work** [hyperledger-archives/fabric#520]
# Summary:

Invoke transaction is not propagated to all the peers when chaincode is invoked on one peer
# Observed In:

Had a similar bug before. Current test is with Git pull request - https://github.com/openblockchain/obc-peer/pull/519
# Priority:

**Highest**
# Description:
## Abstract:

The objective of this test is to deploy, invoke & query the example chaincode. This issue is seen on a multi-peer network only. Rest APIs are used to deploy, invoke and query.
### openchain.yaml:

Attached at the end of the report.
### Environment variables set:

`OPENCHAIN_PEER_ADDRESS`, `OPENCHAIN_VM_DOCKER_TLS_ENABLED`, `OPENCHAIN_VM_DOCKER_TLS_CERT_FILE`, `OPENCHAIN_VM_DOCKER_TLS_CA_FILE`, `OPENCHAIN_VM_DOCKER_TLS_KEY_FILE`, `OPENCHAIN_VM_ENDPOINT`, `OPENCHAIN_PEER_ID`, `OPENCHAIN_PEER_ADDRESSAUTODETECT=false`, `OPENCHAIN_PEER_LISTENADDRESS`
## Scenario:
1. Setup the network with 5 peers as mentioned above.
2. Call _deploy_ api with the following configuration:
   
   URL: `http://<peer_ip>:<peer_port>/devops/deploy`
   METHOD: `POST`
   BODY: 
   
   ```
   {
   "type": "GOLANG",
   "chaincodeID": {
     "path": "https://hub.jazz.net/git/averyd/cc_ex02/chaincode_example02"
   },
   "ctorMsg": {
     "function": "init",
     "args": [
       "a", "100", "b", "200"
     ]
   }
   } 
   ```
3. Call _query_ api with the following configuration on all peers:
   URL: `http://<peer_ip>:<peer_port>/devops/query`
   METHOD: `POST`
   BODY: 
   
   ```
   {
   "chaincodeSpec": {
   "type": "GOLANG",
   "chaincodeID": {
     "name": "09c08b4770c13cbbd877eb96728a4644068f77be1b85489f9b51700c046429af214d43c8d4428422f7a5c785d5afdc3837e9e86abaab3da1f5786d353418d307"
   },
   "ctorMsg": {
     "function": "query",
     "args": [
       "a"
     ]
   }
   }
   } 
   ```
4. Call _invoke_ api with the following configuration on any one peer:
   URL: `http://<peer_ip>:<peer_port>/devops/invoke`
   METHOD: `POST`
   BODY: 
   
   ```
   {
   "chaincodeSpec": {
   "type": "GOLANG",
   "chaincodeID": {
     "name": "09c08b4770c13cbbd877eb96728a4644068f77be1b85489f9b51700c046429af214d43c8d4428422f7a5c785d5afdc3837e9e86abaab3da1f5786d353418d307"
   },
   "ctorMsg": {
     "function": "invoke",
     "args": [
       "a", "b", "20"
     ]
   }
   }
   }
   ```
5. Call _query_ api on all peers with the configuration mentioned in 3.
### Behavior:

  The peer on which the chaincode is invoked has the correct value. The other peers may or may not have the correct values.
### Expected Behavior:

  All the peers should return the same value for query api on the same variable in step 3 & step 5.
## Debug:

  Logs were observed:

  Peer which was invoked showed the logs with following text - sending message to the other peers:

```
[36m00:26:04.952 [peer] SendMessage -> DEBU 17dc[0m Sending message to stream of type: CONSENSUS
[36m00:26:05.725 [peer] SendMessage -> DEBU 17e1[0m Sending message to stream of type: CONSENSUS 
[36m00:26:06.526 [peer] SendMessage -> DEBU 17e2[0m Sending message to stream of type: CONSENSUS 
[36m00:26:07.323 [peer] SendMessage -> DEBU 17e3[0m Sending message to stream of type: CONSENSUS 
```

  None of the peers or only a few of the peers show the following text in the log:

`RecvMsg -> DEBU 17e1[0m Handling OpenchainMessage of type: CONSENSUS`

So, I assume the remaining peers never received the consensus message.
## Note:

  Previously there was a similar issue where some of the peers would not receive the invoke transaction update. Now, none of the peers receive it.
## openchain.yaml

```

---
###############################################################################
#
#    CLI section
#
###############################################################################
cli:

    # The address that the cli process will use for callbacks from chaincodes
    address: 0.0.0.0:30304



###############################################################################
#
#    REST section
#
###############################################################################
rest:

    # Enable/disable setting for the REST service.
    enabled: true

    # The address that the REST service will listen on for incoming requests.
    address: 0.0.0.0:5000




###############################################################################
#
#    Peer section
#
###############################################################################
peer:

    # Peer Version following version semantics as described here http://semver.org/
    # The Peer supplies this version in communications with other Peers
    version:  0.1.0

    # The Peer id is used for identifying this Peer instance.
    id: jdoe

    # The privateKey to be used by this peer
    privateKey: 794ef087680e2494fa4918fd8fb80fb284b50b57d321a31423fe42b9ccf6216047cea0b66fe8365a8e3f2a8140c6866cc45852e63124668bee1daa9c97da0c2a

    # The networkId allows for logical seperation of networks
    # networkId: dev
    # networkId: test
    networkId: dev

    Dockerfile:  |
        from golang:1.5.1
        ENV GO15VENDOREXPERIMENT=1
        # Install RocksDB
        RUN cd /opt  && git clone --branch v4.1 --single-branch --depth 1 https://github.com/facebook/rocksdb.git && cd rocksdb
        WORKDIR /opt/rocksdb
        RUN make shared_lib
        ENV LD_LIBRARY_PATH=/opt/rocksdb:$LD_LIBRARY_PATH
        RUN apt-get update && apt-get install -y libsnappy-dev zlib1g-dev libbz2-dev
        # Copy GOPATH src and install Peer
        COPY src $GOPATH/src
        RUN mkdir -p /var/openchain/db
        WORKDIR $GOPATH/src/github.com/openblockchain/obc-peer/
        RUN CGO_CFLAGS="-I/opt/rocksdb/include" CGO_LDFLAGS="-L/opt/rocksdb -lrocksdb -lstdc++ -lm -lz -lbz2 -lsnappy" go install && cp $GOPATH/src/github.com/openblockchain/obc-peer/openchain.yaml $GOPATH/bin

    # The Address this Peer will listen on
    listenAddress: 0.0.0.0:30303
    # The Address this Peer will bind to for providing services
    address: 0.0.0.0:30303
    # Whether the Peer should programmatically determine the address to bind to. This case is useful for docker containers.
    addressAutoDetect: false


    # Logging settings
    logging:
        # Logging level, can be one of [error|warning|info|debug]
        # One of: CRITICAL | ERROR | WARNING | NOTICE | INFO | DEBUG
        level: DEBUG

    # Peer port to accept connections on
    port:    30303
    # Peer's setting for GOMAXPROCS
    gomaxprocs: 2
    workers: 2

    # Sync related configuration
    sync:
        blocks:
            # Channel size for readonly SyncBlocks messages channel for receiving blocks from oppositie Peer Endpoints.
            # NOTE: currently messages are not stored and forwarded, but rather lost if the channel write blocks.
            channelSize: 1
        state:
            snapshot:
                # Channel size for readonly syncStateSnapshot messages channel for receiving state deltas for snapshot from oppositie Peer Endpoints.
                # NOTE: currently messages are not stored and forwarded, but rather lost if the channel write blocks.
                channelSize: 50
            deltas:
                # Channel size for readonly syncStateDeltas messages channel for receiving state deltas for a syncBlockRange from oppositie Peer Endpoints.
                # NOTE: currently messages are not stored and forwarded, but rather lost if the channel write blocks.
                channelSize: 5

    # Validator defines whether this peer is a validating peer or not, and if
    # it is enabled, what consensus plugin to load
    validator:
        enabled: true

        # Consensus plugin to use. The value is the name of the plugin, e.g. obcpbft, noops
        consensus: noops

         # List of validating peers.
        replicas:
            # For MVP, assume list is static.
            ips: 172.17.0.2:30303 172.17.0.3:30303 172.17.0.4:30303 172.17.0.5:30303
            # Base64-encoded values.
            hashes: a, b, c, d

        events:
            # The address that the Event service will be enabled on the validator
            address: 0.0.0.0:31315

            # total number of events that could be buffered without blocking the validator sends
            buffersize: 100

            # milliseconds timeout for producer to send an event.
            # if < 0, if buffer full, unblocks immediately and not send
            # if 0, if buffer full, will block and guarantee the event will be sent out
            # if > 0, if buffer full, blocks till timeout
            timeout: 10

    # TLS Settings for p2p communications
    tls:
        enabled:  false
        cert:
            file: testdata/server1.pem
        key:
            file: testdata/server1.key
        # The server name use to verify the hostname returned by TLS handshake
        server-host-override:

    # PKI member services properties
    pki:
        eca:
            paddr: localhost:50051
        tca:
            paddr: localhost:50051
        tlsca:
            paddr: localhost:50051

    # Peer discovery settings.  Controls how this peer discovers other peers
    discovery:

        # The root nodes are used for bootstrapping purposes, and generally supplied through ENV variables
        rootnode:

        # The duration of time between attempts to asks peers for their connected peers
        period:  5s

        ## leaving this in for example of sub map entry
        # testNodes:
        #    - node   : 1
        #      ip     : 127.0.0.1
        #      port   : 30303
        #    - node   : 2
        #      ip     : 127.0.0.1
        #      port   : 30303

        # Should the discovered nodes and their reputations
        # be stored in DB and persisted between restarts
        persist:    true

        # if peer discovery is off
        # the peer window will show
        # only what retrieved by active
        # peer [true/false]
        enabled:    true

        # number of workers that
        # tastes the peers for being
        # online [1..10]
        workers: 8

        # the period in seconds with which the discovery
        # tries to reconnect to successful nodes
        #0 means the nodes are not reconnected
        touchPeriod: 600

        # the maximum nuber of nodes to reconnect to
        # -1 for unlimited
        touchMaxNodes: 100

    # Path on the file system where peer will store data
    fileSystemPath: /var/openchain/production

### NOTE: The validator section below is not needed and will be removed - BN
###############################################################################
#
#    Validator section
#
###############################################################################
validator:
    enabled: false
    address: 0.0.0.0:30304
    # TLS Settings for p2p communications
    tls:
        enabled:  false
        cert:
            file: testdata/server1.pem
        key:
            file: testdata/server1.key
        # The server name use to verify the hostname returned by TLS handshake
        server-host-override:
    # Peer discovery settings.  Controls how this peer discovers other peers
    discovery:

        # The root nodes are used for bootstrapping purposes, and generally supplied through ENV variables
        rootnode:

    validity-period: 
        verification: false

###############################################################################
#
#    VM section
#
###############################################################################
vm:

    # Endpoint of the vm management system.  For docker can be one of the following in general
    # unix:///var/run/docker.sock
    # http://localhost:4243
    # tcp://localhost:2376
    endpoint: unix:///var/run/docker.sock

    # settings for docker vms
    docker:
        tls:
            enabled: false
            cert:
                file: /path/to/server.pem
            ca:
                file: /path/to/ca.pem
            key:
                file: /path/to/server-key.pem

###############################################################################
#
#    Chaincode section
#
###############################################################################
chaincode:

    # The id is used by the Chaincode stub to register the executing Chaincode ID with the Peerand is generally supplied through ENV variables
    # the Path form of ID is provided when deploying the chaincode. The name is used for all other requests. The name is really a hashcode
    # returned by the system in response to the deploy transaction. In development mode where user runs the chaincode, the name can be
    # any string
    id:
        path:
        name:

    golang:

        # This is the basis for the Golang Dockerfile.  Additional commands will be appended depedendent upon the chaincode specification.
        Dockerfile:  |
            from golang:1.5.1
            ENV GO15VENDOREXPERIMENT=1
            COPY src $GOPATH/src
            WORKDIR $GOPATH

    #timeout in millisecs for starting up a container and waiting for Register to come through. 1sec should be plenty for chaincode unit tests
    startuptimeout: 1000

    #timeout in millisecs for deploying chaincode from a remote repository.
    deploytimeout: 30000

    #mode - options are "dev", "net"
    #dev - in dev mode, user runs the chaincode after starting validator from command line on local machine
    #net - in net mode validator will run chaincode in a docker container

    mode: net

    installpath: /go/bin/

###############################################################################
#
#    Ledger section - ledger configuration encompases both the blockchain
#    and the state
#
###############################################################################
ledger:

  blockchain:

    # Define the genesis block
    genesisBlock:

      # Deploy chaincodes into the genesis block
      # chaincode:
      #     path: github.com/openblockchain/obc-peer/openchain/example/chaincode/chaincode_example01
      #     type: GOLANG
      #     constructor:
      #       func: init
      #       args:
      #         - alice
      #         - "4"
      #         - bob
      #         - "10"

    state:

      # Control the number state deltas that are maintained. This takes additional
      # disk space, but allow the state to be rolled backwards and forwards
      # without the need to replay transactions.
      deltaHistorySize: 500

    # The data structure in which the state will be stored. Different data
    # structures may offer different performance characteristics. Options are
    # 'buckettree' and 'trie'. If not set, the default data structure is the
    # 'buckettree'. This CANNOT be changed after the DB has been created.
    dataStructure: buckettree

    deploy-system-chaincode: true
###############################################################################
#
#    Security section -
#
###############################################################################
security:
    enabled: false
    # To enroll NVP or VP with Member Services PKI. These parameters are for
    #1 time use. They will not be valid the subsequent times without un-enroll
    enrollID: donq
    enrollSecret: 7avZQLwcUe9q
    # To enable privacy and confidentiality of transactions (requires security to be enabled)
    privacy: false

```

**Comment by muralisrini** [comment]
Able to recreate something like this on a dev environment. Assign to self for now

**Comment by muralisrini** [comment]
can you try this? ... in addition to OPENCHAIN_PEER_ADDRESS and similar env vars you pass to the docker run command when starting peer, can you add OPENCHAIN_LEDGER_STATE_DELTAHISTORYSIZE=500 and try ?

**Comment by srderson** [comment]
@muralisrini I'll assume the problem is the configuration key mismatch we discussed tonight. I'll assign to myself and resolve tomorrow.

**Comment by srderson** [comment]
Bug was introduced in commit 7754c1ff986d5477c868aecc4dc395fb304ed3f1
The state key was moved under blockchain.

**Comment by mrshah-at-work** [comment]
@muralisrini I tried OPENCHAIN_LEDGER_STATE_DELTAHISTORYSIZE=500, not helpful.
See attached image there was an invoke ["a", "b", "20"] for example 02 between the 2 reads in the snapshot.

![image](https://cloud.githubusercontent.com/assets/15948008/12451117/7f46ac00-bf55-11e5-830c-28416ec95ab8.png)

**Comment by muralisrini** [comment]
There might be several issues. Without the OPENCHAIN_LEDGER_STATE_DELTAHISTORYSIZE you should not get more than 1 successful invoke. Now that we are past that, will continue to look

**Comment by muralisrini** [comment]
assigned back to self to continue with debug/fix.

**Comment by muralisrini** [comment]
fixed via PR #544

**Comment by muralisrini** [comment]
The next issue appears to be due to a grpc stream buffering block (as confirmed by experiments by @mrshah-at-ibm ) on the network feedback loop between peers 
- peer loop
  - recv
  -  exec transaction
  -  broadcast SYNC_BLOCK_ADDED to peers 

If block is large, broadcast from peer A .. peer X would have to wait for it to be drained (via recv) however  the broadcast might chain back to a peer waiting for A to drain (typically  SYNC_BLOCK_ADDED but it could be some other message as well). 

If this is confirmed, we'll have to look at this side-by-side with #529

**Comment by muralisrini** [comment]
Ran successfully an overnight test breaking the recv/send loop above (but everything else in the 5-peer scenario remaining same). The above theory is likely correct.

**Comment by binhn** [comment]
@muralisrini good find!
We can attack SYNC_BLOCK_ADDED by 
1) not sending it to VPs
2) remove payload from TXs  
3) deliver it on a separate channel

**Comment by jeffgarratt** [comment]
@binhn @mrshah-at-ibm @srderson @muralisrini Spoke with @mrshah-at-ibm and @muralisrini and agreed that this is in general not an issue that should be surfaced unless we exceed a TBD threshold on message size wrt GRPC.  We agreed to investigate (and produce tests) to determine behavioral patterns wrt GRPC and buffers.  In general, this issue should NOT be critical at the moment.  If there is disagreement, please comment.   There is a committed behavioral test that reproduces @mrshah-at-ibm 's test case with appropriately constructed chaincode deployments (ie. not overly large) and it passes.

**Comment by muralisrini** [comment]
@jeffgarratt ... It could happen with small messages too (and I believe it did in @mrshah-at-ibm env). However, I  I think I can confirm (pending outcome of a test - being ultra cautious given the nature of this beast  :-) ) the network deadlock scenario outlined above.  Trying a quick test without the lock to test the theory.... once confirmed we can work on the correct fix.

**Comment by muralisrini** [comment]
@jeffgarratt : can you take a look at this potential fix please https://github.com/muralisrini/obc-peer/tree/dont_lock_in_read_write_loop%23520

It basically does two things
- clones map to iterate over to avoid locking across SendMessage
- nils Transaction payload in SYNC_BLOCK_ADDED message in noops 

The lock is the main fix. Note while this is the only handlerMap Lock across SendMessages there are other Lock (e.g. syncBlocksMutex Lock) that could raise similar issues.

**Comment by binhn** [comment]
@muralisrini if you create a in-progress pull request, we can easily see the changes

**Comment by muralisrini** [comment]
@binh I will do that within the hour. I want to put in one more change for consideration,

**Comment by muralisrini** [comment]
PR #576 has been running with a clone of @mrshah-at-ibm env for many hours successfully.

**Comment by binhn** [comment]
Fix has been delivered yesterday commit 1e08f4b767d7b84425c80916cddbe94afbc075b4

**Comment by mrshah-at-work** [comment]
Closing the issue, fixed as per above.
